jbi/queue.py

```diff
@@ -0,0 +1,258 @@
+        logger.debug(
+            "Wrote item %s for bug %s to path %s",
+            item.identifier,
+            item.payload.bug.id,
+            path,
+        )
+        logger.debug("%d items in dead letter queue", await self.size())
+
+    async def remove(self, bug_id: int, identifier: str):
+        bug_dir = self.location / f"{bug_id}"
+        item_path = bug_dir / (identifier + ".json")
+        item_path.unlink(missing_ok=True)
+        logger.debug("Removed %s from queue for bug %s", identifier, bug_id)
+        if not any(bug_dir.iterdir()):
+            bug_dir.rmdir()
+            logger.debug("Removed directory for bug %s", bug_id)
+
+    async def get(self, bug_id: int) -> AsyncIterator[QueueItem]:
+        folder = self.location / str(bug_id)
+        if not folder.is_dir():
+            return
+            yield
+        for path in sorted(folder.iterdir()):
+            yield QueueItem.parse_file(path)
+
+    async def get_all(self) -> dict[int, AsyncIterator[QueueItem]]:
+        all_items: dict[int, AsyncIterator[QueueItem]] = {}
+        for filesystem_object in self.location.iterdir():
+            if filesystem_object.is_dir():
+                all_items[int(filesystem_object.name)] = self.get(filesystem_object)
+        return all_items
+
+    async def size(self, bug_id=None) -> int:
+        location = self.location / str(bug_id) if bug_id else self.location
+        return sum(1 for _ in location.rglob("*.json"))
+
+
+class InvalidQueueDSNError(Exception):
+    pass
+
+
+class DeadLetterQueue:
+    backend: QueueBackend
+
+    def __init__(self, dsn: FileUrl | str | ParseResult):
+        dsn = urlparse(url=dsn) if isinstance(dsn, str) else dsn
+
+        if dsn.scheme != "file":
+            raise InvalidQueueDSNError(f"{dsn.scheme} is not supported")
+        self.backend = FileBackend(dsn.path)
+
+    def ready(self) -> list[dockerflow.checks.CheckMessage]:
+        """Heartbeat check to assert we can write items to queue
+
+        TODO: Convert to an async method when Dockerflow's FastAPI integration
+        can run check asynchronously
+        """
+
+        ping_result = self.backend.ping()
+        if ping_result is False:
+            return [
+                dockerflow.checks.Error(
+                    f"queue with f{str(self.backend)} backend unavailable"
+                )
+            ]
+        return []
+
+    async def postpone(self, payload: bugzilla.WebhookRequest) -> None:
+        """
+        Postpone the specified request for later.
+        """
+        item = QueueItem(payload=payload)
+        await self.backend.put(item)
+
+    async def track_failed(
+        self, payload: bugzilla.WebhookRequest, exc: Exception
+    ) -> None:
+        """
+        Store the specified payload and exception information into the queue.
+        """
+        item = QueueItem(
+            payload=payload,
+            error=PythonException.from_exc(exc),
+        )
```

Comment:
Should we be smart here and increment the `retries` field if the request is already in the queue?

Or should we remove the `retries` field for now?

Reply:
Thinking about it more, I'm in favor of removing retries property. We can calculate how many times we've retried a given entry or look at the logs if we need to. Updating it each time will cause more writes to the file share, and increase the risk of a file getting corrupted during a write (although that should be a very low risk).